components/builder-api/src/server/services/events.rs

```diff
@@ -0,0 +1,52 @@
+use std::{thread,
+          time::Duration};
+
+use crate::config::KafkaCfg;
+
+use cloudevents::Event;
+use cloudevents_sdk_rdkafka::{FutureRecordExt,
+                              MessageRecord};
+
+use rdkafka::{config::ClientConfig,
+              producer::{FutureProducer,
+                         FutureRecord}};
+
+pub const KAFKA_DEFAULT_TOPIC_NAME: &str = "bldr_event_init";
+
+pub struct KafkaProducer(pub FutureProducer);
+
+impl KafkaProducer {
+    pub fn new(config: KafkaCfg) -> Self {
+        debug!("Creating new kafka producer, config: {{bootstrap_nodes: {}, client_id: {}}}",
+               config.bootstrap_nodes, config.client_id);
+        loop {
+            match ClientConfig::new().set("bootstrap.servers", &config.bootstrap_nodes)
+                                     .set("message.timeout.ms", &config.message_timeout)
+                                     .set("client.id", &config.client_id)
+                                     .set("sasl.username", &config.api_key)
+                                     .set("sasl.password", &config.api_secret_key)
+                                     .set("security.protocol", "SASL_SSL")
+                                     .set("sasl.mechanisms", "PLAIN")
+                                     .create()
+            {
+                Ok(iprodr) => return KafkaProducer(iprodr),
+                Err(err) => error!("Error initializing kafka producer, will retry: {:?}", err),
+            }
+            thread::sleep(Duration::from_millis(config.connection_retry_ms));
```

Comment:
It might be cleaner to implement `TryFrom<KafkaCfg> for KafkaProducer` here, and handle retries explicitly. That decouples the fundamental creation of the producer from the retry logic, rather than mixing them together.

Also, it would be good to have a limit on the number of retries; otherwise you could just spin indefinitely.

Reply:
That's a pretty great idea. When the feature is enabled i think we _do_ want to retry until a connection is made available.

Reply:
That could absolutely be a valid outcome of this. That particular decision would be really great to see explicitly handled in the service startup logic, though, rather than buried down here. That kind of thing is really in the domain of the overall service as a whole, rather than something fundamental to either of these types (and embedding that logic here strongly implies that it really is in the domain of the types).